File: 39. Combination Sum.go
// https://leetcode.com/problems/combination-sum/

package leetcode

func combinationSum(candidates []int, target int) [][]int {
	result, wc := [][]int{}, []int{}

	var backtrack func(int, int)
	backtrack = func(i, remainder int) {
		switch {
		case i >= len(candidates) || remainder < 0:
			return
		case remainder == 0:
			dupe := make([]int, len(wc))
			copy(dupe, wc)
			result = append(result, dupe)
		default:
			wc = append(wc, candidates[i])
			backtrack(i, remainder-candidates[i])
			wc = wc[:len(wc)-1]
			backtrack(i+1, remainder)
		}
	}

	backtrack(0, target)
	return result
}

File: 39. Combination Sum.js
// https://leetcode.com/problems/combination-sum/

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const [result, wc] = [[], []];

    const backtrack = (i, remainder) => {
        if(i >= candidates.length || remainder < 0) return;
        if(remainder === 0) return result.push(wc.slice());

        const curr = candidates[i];
        wc.push(curr);
        backtrack(i, remainder - curr); // try using this item again
        wc.pop();
        backtrack(i+1, remainder); // try not using this item
    };

    backtrack(0, target);
    return result;
};